### PR TITLE
feat: migrate to v4

### PR DESCRIPTION
**Pull request by ndom91**
Migrated everything that needs changed so far.

I did, however, run into one console error regarding `jwt` that I wasn't sure about. Maybe I forgot to migrate something?

For example, on the "API" page of the example app, which is supposed to print the current session and jwt info, I get the following in my node console:

```bash
event - build page: /api/examples/jwt
TypeError [ERR_INVALID_ARG_TYPE]: The "key" argument must be of type string or an instance of SecretKeyObject, ArrayBuffer, TypedArray, DataView, or Buffer. Received undefined
    at Object.hkdfSync (node:internal/crypto/hkdf:134:7)
    at hkdf (/opt/ndomino/next-auth-typescript-example/node_modules/next-auth/dist/lib/jwt.js:127:40)
    at getDerivedSigningKey (/opt/ndomino/next-auth-typescript-example/node_modules/next-auth/dist/lib/jwt.js:143:18)
    at decode (/opt/ndomino/next-auth-typescript-example/node_modules/next-auth/dist/lib/jwt.js:81:96)
    at getToken (/opt/ndomino/next-auth-typescript-example/node_modules/next-auth/dist/lib/jwt.js:109:12)
    at jwt (webpack-internal:///./pages/api/examples/jwt.ts:11:78)
    at Object.apiResolver (/opt/ndomino/next-auth-typescript-example/node_modules/next/dist/server/api-utils.js:101:15)
    at processTicksAndRejections (node:internal/process/task_queues:94:5)
    at async DevServer.handleApiRequest (/opt/ndomino/next-auth-typescript-example/node_modules/next/dist/server/next-server.js:760:9)
    at async Object.fn (/opt/ndomino/next-auth-typescript-example/node_modules/next/dist/server/next-server.js:651:37) {
  code: 'ERR_INVALID_ARG_TYPE'
}
```

On both the normal and typescript example apps btw.

Any idea whats going on there? Did I miss something in terms of the migration?

Obviously there will probably be a few more small things to change before this is really up to snuff for `v4.0.0` (final), but hopefully this takes care of most of the work and then we can merge this PR whenever the day comes tada

### CODE REVIEW

**Comment by motleydev** [comment]
I attempted to use this PR and defining an email provider as well as installing nodemailer throws an "adapter undefined" error.

Error: `[next-auth][error][EMAIL_REQUIRES_ADAPTER_ERROR] `

**Comment by balazsorban44** [comment]
as the message says, you'll need an adapter. 😅

https://next-auth.js.org/adapters/overview

@ndom91 we could explain it a bit better in:
https://next-auth.js.org/errors#email_requires_adapter_error

**Comment by motleydev** [comment]
> as the message says, you'll need an adapter. 😅
> 
> https://next-auth.js.org/adapters/overview
> 
> @ndom91 we could explain it a bit better in:
> https://next-auth.js.org/errors#email_requires_adapter_error

Hah, I did get THAT far. 😂  I installed nodemailer like mentioned above, which is what the docs indicated as the only required "adapter" needed - am I missing something additional to that? Sorry for the noob Q - hopefully this will help clarify some extra points for those that come after. :)

**Comment by balazsorban44** [comment]
nodemailer is not an adapter, it is a library that helps you sending emails. an adapter is a way for us to save data in a database. which is a requirement for the Email provider. see the docs link I sent you.

**Comment by motleydev** [comment]
> nodemailer is not an adapter, it is a library that helps you sending emails. an adapter is a way for us to save data in a database. which is a requirement for the Email provider. see the docs link in my previous reply

@balazsorban44 ah - I see now. From the looks of it, then, there's no longer a native postgres adapter? Previously installing pg and identifying the connecting string was what was required. Just to make sure I understand before writing a custom adapter. Thanks.

**Comment by balazsorban44** [comment]
We don't ship with an adapter anymore, because it made the bundle unnecessarily massive for those who did not use any.

Make sure you read the whole migration docs! https://next-auth.js.org/getting-started/upgrade-v4#1-adapters

**Comment by ndom91** [comment]
@motleydev yeah so when using the email provider you are required to use a real database adapter because we need to store the email magic token thing somewhere longer term to verify it whenever the user clicks the email link. 

You can think of nodemailer as just a dependency for the email provider, not an adapter in the classical next auth sense. Adapters for us are strictly the database adapters. 

Well definitely make this clearer in the docs!

**Comment by balazsorban44** [comment]
The current stable version is v3, so this PR will be delayed until v4 becomes the stable one. I still wish to have the ability to do minor breaking changes while in the beta phase.

When someone creates an example from this template, I want them to use v3.

**Review by balazsorban44 — CHANGES_REQUESTED** [review]
Could you update this to use `beta.7` To be ready for merge?

**Comment by ndom91** [comment]
@balazsorban44 updated the PR to `beta.7` :+1:

**Comment by ndom91** [comment]
@balazsorban44 this is seemingly having those typescript issues resolving `next-auth/react` again at build time on vercel (https://vercel.com/nextauthjs/next-auth-example/3PyFJEmnfyyF5eMZyjjG93jCUHXj)

File: package.json [hunk 1]
```diff
@@ -16,12 +16,14 @@
...
-    "Balázs Orbán <info@balazsorban.com>"
+    "Balázs Orbán <info@balazsorban.com>",
+    "Nico Domino <yo@ndo.dev>"
   ],
   "license": "ISC",
   "dependencies": {
-    "next": "^11.0.0",
-    "next-auth": "latest",
+    "next": "^12.0.1",
+    "next-auth": "^4.0.0-beta.7",
```
Comment (balazsorban44) [inline package.json:25]:
```suggestion
    "next-auth": "4.0.0-beta.7",
```

Use either exact version, or `beta` for now.

Comment (ndom91) [reply package.json]: Good catch!

**Comment by balazsorban44** [comment]
image
we will need to comment out the email adapter

**Comment by ndom91** [comment]
> image we will need to comment out the email adapter

Okay, done 👍

**Pull request merged**